### PR TITLE
feat: delete provider & resources

### LINKED ISSUE #154: Add a method to delete providers and its corresponding resources
Request for a feature to delete providers and its corresponding resources from guardian. 

Currently, if a provider was added to guardian, then guardian automatically fetches the resources and populates the resources table. If in case, the provider and its resources should not be there in guardian at all and there are no appeals tied up with this provider, the only way to remove them would be to run a query on the guardian db directly since there are no api's implemented in guardian for removing providers.

### PR DESCRIPTION
**Pull request by rahmatrhd**
closes #154 
proton PR: https://github.com/odpf/proton/pull/140

features:
- delete single resource
- delete single provider + the corresponding resources

### CODE REVIEW

File: core/provider/service.go [hunk 1]
```diff
@@ -276,6 +277,31 @@ func (s *Service) RevokeAccess(a *domain.Appeal) error {
...
+		"provider_urn":  p.URN,
+	})
+	if err != nil {
+		return fmt.Errorf("retrieving related resources: %w", err)
+	}
+	var resourceIds []string
+	for _, r := range resources {
+		resourceIds = append(resourceIds, r.ID)
+	}
+	// TODO: execute in transaction
```
Comment (mabdh) [inline core/provider/service.go:297]: does it mean we do this in another PR?
or this one is missing in current PR?

File: api/handler/v1beta1/grpc.go [hunk 2]
```diff
@@ -186,6 +187,17 @@ func (s *GRPCServer) UpdateProvider(ctx context.Context, req *guardianv1beta1.Up
 	}, nil
 }
 
+func (s *GRPCServer) DeleteProvider(ctx context.Context, req *guardianv1beta1.DeleteProviderRequest) (*emptypb.Empty, error) {
```
[the remark this answers is not in the document]
Comment (rahmatrhd) [reply api/handler/v1beta1/grpc.go:190]: I was following example in here https://cloud.google.com/apis/design/standard_methods#delete, and `error` is required by grpc-gateway CMIIW

File: core/provider/service.go [hunk 3]
```diff
@@ -276,6 +277,31 @@ func (s *Service) RevokeAccess(a *domain.Appeal) error {
...
+		return fmt.Errorf("getting provider details: %w", err)
+	}
+
+	resources, err := s.resourceService.Find(map[string]interface{}{
+		"provider_type": p.Type,
+		"provider_urn":  p.URN,
+	})
+	if err != nil {
+		return fmt.Errorf("retrieving related resources: %w", err)
+	}
```
[the remark this answers is not in the document]
Comment (rahmatrhd) [reply core/provider/service.go:292]: will handle this in grpc handler

File: core/provider/service.go (hunk 1 above)
Comment (rahmatrhd) [reply core/provider/service.go:297]: won't do in this PR, since there's also a big changes here https://github.com/odpf/guardian/pull/155, will do at least after both got merged

**Comment by rahmatrhd** [comment]
@mabdh added the proton PR link

File: core/provider/service.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (rahmatrhd) [reply core/provider/service.go:292]: just check that resource.Find won't return resouce.Notfound since it's a List method, not GetOne. Therefore, if the returned resources are empty, that's fine, shouldn't be considered as resource not found

**Review by mabdh — APPROVED** [review]

**Review by StewartJingga — APPROVED** [review]
